### DISCUSSION TITLE
Adding cmp source problem

### DISCUSSION THREAD

**Discussion by bassamsdata** [LazyVim/LazyVim#202] [Q&A] [answered]
I'm trying to add cmp source to nvim-cmp but I couldn't succeed on that.
I'm using the code in the `example.lua` to add `cmp_nvim_r` from the plugin [cmp-nvim-r](https://github.com/jalvesaq/cmp-nvim-r)
`opts.sources = cmp.config.sources(vim.list_extend(opts.sources, { { name = "cmp_nvim_r" } }))`
there is a diagnosis hint that this function needs 4 arguments but I'm not sure if this the problem, since the function has 2 defaults.
but it didn't work. I add it to the super tab snippet but it didn't work either:
```r
  {
    "hrsh7th/nvim-cmp",
    dependencies = {
      "hrsh7th/cmp-emoji",
    },
    ---@param opts cmp.ConfigSchema
    opts = function(_, opts)
      local has_words_before = function()
        unpack = unpack or table.unpack
        local line, col = unpack(vim.api.nvim_win_get_cursor(0))
        return col ~= 0 and vim.api.nvim_buf_get_lines(0, line - 1, line, true)[1]:sub(col, col):match("%s") == nil
      end

      local luasnip = require("luasnip")
      local cmp = require("cmp")
      opts.sources = cmp.config.sources(vim.list_extend(opts.sources, { { name = "cmp_nvim_r" } }))

      opts.mapping = vim.tbl_extend("force", opts.mapping, {
        ["<Tab>"] = cmp.mapping(function(fallback)
          if cmp.visible() then
            cmp.select_next_item()
            -- You could replace the expand_or_jumpable() calls with expand_or_locally_jumpable()
            -- they way you will only jump inside the snippet region
          elseif luasnip.expand_or_jumpable() then
            luasnip.expand_or_jump()
          elseif has_words_before() then
            cmp.complete()
          else
            fallback()
          end
        end, { "i", "s" }),
        ["<S-Tab>"] = cmp.mapping(function(fallback)
          if cmp.visible() then
            cmp.select_prev_item()
          elseif luasnip.jumpable(-1) then
            luasnip.jump(-1)
          else
            fallback()
          end
        end, { "i", "s" }),
      })
    end,
  },
```

Thank you so much for your help

**Comment by folke** [comment, marked as answer]
You did not add the source here? The dependncy list cmp-emoji, which is not correct

**Reply by bassamsdata** [reply]
sorry, but I'm not sure I understood what do you mean, if I add this snippet after installing the plugin `jalvesaq/cmp-nvim-r`, the cmp didn't add the plugin to work, even if I add like dependency which I suspect it's not true to do :(

  ```
{
    "hrsh7th/nvim-cmp",
    opts = function(_, opts)
      local cmp = require("cmp")
      opts.sources = cmp.config.sources(vim.list_extend(opts.sources, { { name = "cmp_nvim_r" } }))
    end,
  }
```
<img width="1342" alt="Screen Shot 2023-02-07 at 11 47 25 AM" src="https://user-images.githubusercontent.com/105807570/217309240-d05e72ee-e337-4c7c-b3bb-7023e1d43a0e.png">

edited: the correct image

**Reply by folke** [reply]
The way you configured it is correct. Not an R user, so not sure why that plugin would not be working.

What does `:CmpStatus` say?

**Reply by bassamsdata** [reply]
thanks for the reply,  the `:CmpStatus` say:
<img width="486" alt="Screen Shot 2023-02-07 at 2 37 57 PM" src="https://user-images.githubusercontent.com/105807570/217349866-86bb21d2-4fa7-4df5-8e0b-12f660f9cdb2.png">

I found the issue which is weird.
Below is the config for the plugin, if I remove the option `ft = { "r", "rmd", "quarto" }` everything work, but once I active the lazy loading option, it doesn't work. it's really weird even though it shows that the `cmp_nvim_r` is ready as shown above (when the ft option is active).

```lua
  {
    "jalvesaq/cmp-nvim-r",
    ft = { "r", "rmd", "quarto" },
    config = function()
      require("cmp_nvim_r").setup({})
    end,
  },
```
Another thing (which is probably a little bug):
If I activated the the `super tab` example snippet, with transparent colorscheme (I'm using `tokyonight` or `rose-pine`), the cmp menu appear in black color :(

https://user-images.githubusercontent.com/105807570/217351515-88298e47-d2b0-4817-95dc-48bade7e6502.mov


https://user-images.githubusercontent.com/105807570/217351654-832a19c9-1bb9-4a93-9892-70765e553391.mov

**Reply by folke** [reply]
Not sure about that black cmp doc window. You can try changing the cmp `window.documentation.winhighlight` Check the cmp docs

**Reply by qizidog** [reply]
I had trouble with the background of documentation window too. It keep not transparent even after setting `winhighlight` in cmp config.
```lua
opts.window = {
    documentation = {
        winhighlight = "Normal:NormalFloat,FloatBorder:FloatBorder,CursorLine:Visual,Search:None",
    },
}
```
However, I fix this problem by set some option in `options.lua`.
```lua
vim.o.pumblend = 0
vim.o.winblend = 0
```
check this issue catppuccin/nvim#412.

**Reply by qizidog** [reply]
@folke 

I checked the help of `pumblend ` but then I get confused. It is said that, 

>  `'pumblend'`  `'pb'` 		number	(default 0)
>  			global
>  	Enables pseudo-transparency for the |popup-menu|. Valid values are in
>  	the range of 0 for fully opaque popupmenu (disabled) to 100 for fully
>  	transparent background. Values between 0-30 are typically most useful.
> 
>  	It is possible to override the level for individual highlights within
>  	the popupmenu using |highlight-blend|. For instance, to enable
>  	transparency but force the current selected element to be fully opaque: >
> 
>  		:set pumblend=15
>  		:hi PmenuSel blend=0

In the LazyVim default config `opt.pumblend = 10 -- Popup blend` is set. However, only when `pumblend` is set to 0 can the background be made transparent, which is the opposite of the description in the help document.

What is the reason for this?

**Reply by folke** [reply]
setting it to 0 will not blend and use your bg color, meaning transparent. setting it to anything else, will blend it. blending with transparent makes no sense, so I assume Neovim then picks a default color
